src/java_bytecode/java_bytecode_language.cpp

```diff
@@ -256,6 +257,106 @@ static void infer_opaque_type_fields(
   }
 }
 
+/// Create if necessary and then return a reference to a symbol for a class
+/// literal (i.e. the result of a Classname.class expression in Java). A
+/// constant global symbol will be created if it doesn't already exist.
+/// \param class_ref: the raw bytecode expression for the class literal
+/// \param symbol_table: global symbol table; may gain a new symbol
+/// \return symbol expression for the Class literal.
+static symbol_exprt get_or_create_class_literal_symbol(
+  const exprt &class_ref,
+  symbol_tablet &symbol_table)
+{
+  const irep_idt &class_id = class_ref.type().get(ID_identifier);
+  symbol_typet java_lang_Class("java::java.lang.Class");
+  symbol_exprt symbol_expr(
+    id2string(class_id)+"@class_model",
+    java_lang_Class);
+
+  if(!symbol_table.has_symbol(symbol_expr.get_identifier()))
+  {
+    symbolt new_class_symbol;
+    new_class_symbol.name = symbol_expr.get_identifier();
+    new_class_symbol.type = symbol_expr.type();
+    // name without "java::" prefix:
+    INVARIANT(
+      has_prefix(id2string(new_class_symbol.name), "java::"),
+      "class name should have java:: prefix");
+    new_class_symbol.base_name =
+      id2string(new_class_symbol.name).substr(6);
+    new_class_symbol.mode = ID_java;
+    new_class_symbol.is_lvalue = true;
+    new_class_symbol.is_state_var = true;
+    new_class_symbol.is_static_lifetime = true;
+    symbol_table.add(new_class_symbol);
+  }
+
+  return symbol_expr;
+}
+
+/// Creates global variables for constants mentioned in a given method. These
+/// are either string literals, or class literals (the java.lang.Class instance
+/// returned by `(some_reference_typed_expression).class`). The method parse
+/// tree is rewritten to directly reference these globals.
+/// \param parse_tree: parse tree to search for constant global references
+/// \param symbol_table: global symbol table, to which constant globals will be
+///   added.
+/// \param string_refinement_enabled: true if `--refine-stings` is active,
+///   which changes how string literals are structured.
+static void generate_constant_global_variables(
+  java_bytecode_parse_treet &parse_tree,
+  symbol_tablet &symbol_table,
+  bool string_refinement_enabled)
+{
+  for(auto &method : parse_tree.parsed_class.methods)
+  {
+    for(auto &instruction : method.instructions)
+    {
+      if(instruction.statement == "ldc" ||
+         instruction.statement == "ldc2" ||
+         instruction.statement == "ldc_w" ||
+         instruction.statement == "ldc2_w")
```

Comment:
It'd be great if you could pull this condition into a function. I'm not sure what any of those magic values mean.

Reply:
Added a comment. No other Java bytecode opcodes have such functions, so I'll leave these ones plain as well for now.